get_fm_data.py

```diff
@@ -0,0 +1,200 @@
+            data_point['time'] = datetime.fromtimestamp(price['event_start']/1000).isoformat()
+            data_point['price'] = round((price['event_value'] * conversion) + markup, 2)
+            if data_point['price'] < 0:
+                has_negative_prices = True
+            epex_price_points.append(data_point)
+
+        # To make sure HA considers this as new info a datetime is added
+        new_state = "APEX prices collected at " + now.isoformat()
+        result = {}
+        result['records'] = epex_price_points
+        self.set_state("input_text.epex_prices", state=new_state, attributes=result)
+
+        # FM returns all the prices it has, sometimes it has not retreived new
+        # prices yet, than it communicates the prices it does have.
+        date_latest_price = datetime.fromtimestamp(prices[-1].get('event_start')/1000).isoformat()
+        date_tomorrow = (now + timedelta(days=1)).isoformat()
+        if date_latest_price < date_tomorrow:
+            self.log(f"FM Epex prices seem not renewed yet, latest price at: {date_latest_price}, retry in {delay} minutes.")
+            self.run_in(self.get_epex_prices, delay)
+        else:
+            if has_negative_prices == True:
+                self.notify("Negative electricity prices for tomorrow. Consider"
+                " to check times in the app to optimising electricity usage.")
+            self.attempts_today = 0
+            self.log(f"FM Epex prices succesfully retrieved. Latest price at: {date_latest_price}.")
+        
+        return
+
+
+    def authenticate_with_fm(self):
+        """Authenticate with the FlexMeasures server and store the returned auth token.
+        Hint: the lifetime of the token is limited, so also call this method whenever the server returns a 401 status code.
+        """
+        self.log("Authenticating with FlexMeasures")
+        res = requests.post(
+            self.args["fm_data_api"] + "requestAuthToken",
+            json=dict(
+                email=self.args["fm_data_user_email"],
+                password=self.args["fm_data_user_password"],
+            ),
+        )
+        if not res.status_code == 200:
+            self.log(f"Authentication failed with response {res.json()}")
+        self.fm_token = res.json()["auth_token"]
+
+    def handle_response_errors(self, message, res, description, fnc, *args, **fnc_kwargs):
+        if fnc_kwargs.get("retry_auth_once", True) and res.status_code == 401:
+            self.log(
+                f"Failed to {description} on authorization (possibly the token expired); attempting to reauthenticate once")
+            self.authenticate_with_fm()
+            fnc_kwargs["retry_auth_once"] = False
+            fnc(*args, **fnc_kwargs)
+        else:
+            self.log(f"Failed to {description} (status {res.status_code}): {res} as response to {message}")
+            self.log(f"Failed to {description} (status {res.status_code}): {res.json()} as response to {message}")
```

Comment:
We should refactor so that these functions are defined in one place, for example, as methods of a mixin class, that is subclassed by both this class and `FlexMeasuresClient`.

Reply:
could use some help with this.

Reply:
I deferred this refactoring work to a separate PR.